VR-8149/project creation uses setResources

coryverta opened the pull request:
Uses the CollaboratorService gRPC interface to create workspace role bindings.

coryverta committed:
Implement call to setResources

coryverta committed:
Set ownerID to be the parsed Long user ID

coryverta committed:
Merge branch 'master' into cj/VR-8152/project-creation-uses-setResources

# Conflicts:
#	backend/src/main/java/ai/verta/modeldb/authservice/AuthServiceChannel.java

coryverta committed:
Replace legacy call to create workspace role bindings

Use the new path to UAC instead.

coryverta committed:
Merge branch 'master' into cj/VR-8152/project-creation-uses-setResources

conradoverta commented on backend/src/main/java/ai/verta/modeldb/authservice/PublicRoleServiceUtils.java:
```diff
@@ -266,6 +268,18 @@ public WorkspaceDTO getWorkspaceDTOByWorkspaceId(
   }
 
+  @Override
+  public boolean createWorkspaceRoleBinding(
```
Note: this will create more than just workspace role bindings. It will set owner and visibility too.

coryverta replied:
Easy to rename it now

conradoverta commented on backend/src/main/java/ai/verta/modeldb/authservice/PublicRoleServiceUtils.java:
```diff
@@ -266,6 +268,18 @@ public WorkspaceDTO getWorkspaceDTOByWorkspaceId(
+          Long ownerId,
+          ModelDBServiceResourceTypes resourceType,
+          CollaboratorTypeEnum.CollaboratorType collaboratorType,
+          ProjectVisibility visibility) {
```
Should we use the collaborator visibility?

coryverta replied:
I changed it to use the common VisibilityEnum.Visibility type to make it more generic and maintain OSS compatibility.

conradoverta commented on backend/src/main/java/ai/verta/modeldb/authservice/PublicRoleServiceUtils.java:
```diff
@@ -266,6 +268,18 @@ public WorkspaceDTO getWorkspaceDTOByWorkspaceId(
+  @Override
+  public boolean createWorkspaceRoleBinding(
+          Long workspaceId,
+          WorkspaceType workspaceType,
```
Should we add a method with workspace name? I can see this one being useful for migration, since these are the two fields we have in the db, but during the API call, we'll have just the name. We can let UAC resolve by itself what is the corresponding ID if we send the name.

coryverta replied:
It's an easy overload to add

conradoverta commented on backend/src/main/java/ai/verta/modeldb/authservice/RoleServiceUtils.java:
```diff
@@ -1148,6 +1150,57 @@ public WorkspaceDTO getWorkspaceDTOByWorkspaceId(
+  private ResourceVisibility getResourceVisibility(
+      WorkspaceType workspaceType, ProjectVisibility projectVisibility) {
+    if (workspaceType == WorkspaceType.ORGANIZATION) {
+      if (projectVisibility == ProjectVisibility.ORG_SCOPED_PUBLIC) {
```
We should still be able to create a private project inside an organization.

conradoverta commented on backend/src/main/java/ai/verta/modeldb/authservice/RoleServiceUtils.java:
```diff
@@ -1148,6 +1150,57 @@ public WorkspaceDTO getWorkspaceDTOByWorkspaceId(
+      SetResources setResources =
+          SetResources.newBuilder()
+              .setWorkspaceId(workspaceId)
+              .setOwnerId(ownerId)
```
I believe that UAC will auto-populate the owner, so we don't _need_ to send it here. No harm in sending as long as it's correct though.

coryverta replied:
I made the owner ID optional

conradoverta reviewed (commented):
Minor comments that might be relevant as a follow-up if not easier to do now.

coryverta committed:
Rename createWorkspaceRoleBindings -> createWorkspacePermissions

coryverta committed:
Introduce createWorkspacePermissions override that accepts the name

coryverta committed:
getResourceVisibility support org-private

coryverta committed:
Allow ownerId to be optional

coryverta committed:
Allow projectType to be optional

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge remote-tracking branch 'origin/cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

coryverta committed:
Merge proto changes

coryverta committed:
Migrate to VisibilityEnum.Visibility

coryverta committed:
Test fixes

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

coryverta committed:
Remove Project.SetProjectVisibility

anandjakhaniya committed:
Remove unused code for the creating RoleBinding while insert project on mdb

anandjakhaniya committed:
Merge branch 'cj/VR-8152/project-creation-uses-setResources' of github.com:VertaAI/modeldb into cj/VR-8152/project-creation-uses-setResources

anandjakhaniya commented on backend/src/main/java/ai/verta/modeldb/entities/ProjectEntity.java:
```diff
@@ -73,8 +74,8 @@ public ProjectEntity(Project project) throws InvalidProtocolBufferException {
-  @Column(name = "project_visibility")
-  private Integer project_visibility;
+  @Transient
+  private VisibilityEnum.Visibility projectVisibility = VisibilityEnum.Visibility.PRIVATE;
```
@coryverta we can not remove this directly because while migrating I can not get actual visibility from DB because the entity does not have that field. so in the first part, we have to migrate it and after successfully executing with new visibility logic we will remove it in the second round.

anandjakhaniya replied:
```
 @Column(name = "project_visibility")
  private Integer project_visibility;
```
taking this revert since it is used on the migration

anandjakhaniya committed:
revert old project visibility from ProjectEntity.java for migration use

anandjakhaniya commented on backend/src/main/java/ai/verta/modeldb/project/ProjectServiceImpl.java:
```diff
@@ -1330,47 +1277,6 @@ public void getProjectShortName(
-  public void setProjectVisibility(
-      SetProjectVisibilty request, StreamObserver<SetProjectVisibilty.Response> responseObserver) {
-    try {
-      // Request Parameter Validation
-      if (request.getId().isEmpty()) {
-        String errorMessage = "Project ID not found in SetProjectVisibilty request";
-        LOGGER.info(errorMessage);
-        Status status =
-            Status.newBuilder()
-                .setCode(Code.INVALID_ARGUMENT_VALUE)
-                .setMessage(errorMessage)
-                .addDetails(Any.pack(SetProjectVisibilty.getDefaultInstance()))
-                .build();
-        throw StatusProto.toStatusRuntimeException(status);
-      }
-
-      // Validate if current user has access to the entity or not
-      roleService.validateEntityUserWithUserInfo(
-          ModelDBServiceResourceTypes.PROJECT, request.getId(), ModelDBServiceActions.UPDATE);
-
-      Project project =
-          projectDAO.setProjectVisibility(request.getId(), request.getProjectVisibility());
-      saveAuditLogs(
-          null,
-          ModelDBConstants.UPDATE,
-          Collections.singletonList(project.getId()),
-          String.format(
-              ModelDBConstants.METADATA_JSON_TEMPLATE,
-              "update",
-              "visibility",
-              project.getProjectVisibility().name()));
-      responseObserver.onNext(
-          SetProjectVisibilty.Response.newBuilder().setProject(project).build());
-      responseObserver.onCompleted();
-    } catch (Exception e) {
-      ModelDBUtils.observeError(
-          responseObserver, e, SetProjectVisibilty.Response.getDefaultInstance());
-    }
-  }
-
```
Why this endpoint is removed?? is it intentionally removed?? cc @conradoverta, @coryverta

anandjakhaniya replied:
Ok, i found the Proto change for this https://github.com/VertaAI/protos-all/pull/99/files?file-filters%5B%5D=.proto#diff-05d06fb0d9c3d4b7b133e720c399f054fcea157e127a72581ca7d35cb86949e5L516-L522 so assuming it removed by intention.

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

coryverta committed:
Merge branch 'cj/VR-8152/project-creation-uses-setResources' of github.com:VertaAI/modeldb into cj/VR-8152/project-creation-uses-setResources

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

anandjakhaniya committed:
Fix setProjectWorkspace endpoint (#1758)

Remove setVisibility DAO impl

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Remove ProjectService.setProjectWorkspace

coryverta committed:
Remove ProjectService.setProjectWorkspace

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

Commit:
Update protos from https://github.com/VertaAI/protos-all/pull/99

coryverta committed:
Merge branch 'cj/VR-8152/use-collaborator-visibility' into cj/VR-8152/project-creation-uses-setResources

coryverta committed:
Migrate to using ResourceVisibility for ProjectEntity

coryverta committed:
Migrate to using ResourceVisibility for ProjectEntity

conradoverta commented on backend/src/main/java/ai/verta/modeldb/authservice/RoleServiceUtils.java:
```diff
@@ -961,7 +960,7 @@ private GeneratedMessageV3 getOrgByName(boolean retry, String name) {
     }
 
     if (resourceVisibility != null
-        && (resourceVisibility.equals(ProjectVisibility.PUBLIC)
```
Let's kill this mention of public since it's not in the new enums.

conradoverta commented on backend/src/main/java/ai/verta/modeldb/authservice/RoleServiceUtils.java:
```diff
@@ -1148,8 +1150,96 @@ public WorkspaceDTO getWorkspaceDTOByWorkspaceId(
     }
   }
 
+  private boolean createWorkspacePermissions(
```
I'm pretty confident I already reviewed this method. Is this being moved from somewhere? If yes, what are the changes? It's really hard for me to re-review the same code again.

coryverta replied:
This is a refactor to support using either the workspaceServiceId or the workspaceName.  Those methods now both call this shared private implementation that performs the UAC call.  This function accepts optional parameters for the workspace ID, name, type, and owner, and populates whatever is provided onto the grpc call to UAC.  

This mehtod is called from https://github.com/VertaAI/modeldb/pull/1750/files#diff-c8d019d1d18aa75e6ffdc3d9eee30bdda569c52ea122ddfbdbe326693460ec56R1190-R1198 and https://github.com/VertaAI/modeldb/pull/1750/files#diff-c8d019d1d18aa75e6ffdc3d9eee30bdda569c52ea122ddfbdbe326693460ec56R1210-R1218 below.

conradoverta replied:
Could you highlight what are the differences? I'm pretty sure I'll have a really hard time understanding how it connects to other things due to the size of the method and PR. Having a better idea of what's supposed to have changed will help me focus on that.

coryverta replied:
Sure thing.
https://github.com/VertaAI/modeldb/pull/1750/files#diff-f93a82e7dcaa2d1dbb8859f86aba1d7526b392ac93d4a53896041f414a62d4d2R139-R155 add two new interface methods.  The first is for creating permissions with a `workspaceName`, the second uses the `workspaceServiceId`.

coryverta replied:
In a previous commit, they have nearly identical method bodies.  I consolidated them both to this singular method which takes in both values as Optional and populates them accordingly.

coryverta replied:
https://github.com/VertaAI/modeldb/commit/f077f37e9d078fd0fa5870831e55d55238545031#diff-c8d019d1d18aa75e6ffdc3d9eee30bdda569c52ea122ddfbdbe326693460ec56R1164-R1250

That is the introduction of the refactor, which adds the overload that accepts the workspaceName and then extracts this private method to handle the logic.

coryverta replied:
It all comes down to these lines: https://github.com/VertaAI/modeldb/commit/f077f37e9d078fd0fa5870831e55d55238545031#diff-c8d019d1d18aa75e6ffdc3d9eee30bdda569c52ea122ddfbdbe326693460ec56R1190-R1197

Which check to see if the parameter is provided and add it to the request if so.

conradoverta commented on backend/src/test/java/ai/verta/modeldb/FindProjectEntitiesTest.java:
```diff
@@ -272,7 +274,7 @@ private static void createProjectEntities() {
             .addTags("Tag_7")
             .addTags("Tag_8")
-            .setProjectVisibility(ProjectVisibility.PUBLIC)
+            .setVisibility(ResourceVisibility.PRIVATE)
```
This is going from public to private. I don't know if it breaks anything.

coryverta replied:
For this test is should not matter, since the visibility is not part of the criteria under test.  I'll double check it to be sure.

conradoverta commented on backend/src/main/java/ai/verta/modeldb/utils/RdbmsUtils.java:
```diff
@@ -811,7 +807,7 @@ public static Predicate getValuePredicate(
                 valueExpression,
                 operator,
-                ProjectVisibility.valueOf(value.getStringValue()).ordinal());
+                VisibilityEnum.Visibility.valueOf(value.getStringValue()).ordinal());
```
```suggestion
                ResourceVisibility.valueOf(value.getStringValue()).ordinal());
```
?

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectServiceImpl.java:
```diff
@@ -1484,7 +1387,7 @@ public void findProjects(
 
       ProjectPaginationDTO projectPaginationDTO =
-          projectDAO.findProjects(request, null, userInfo, ProjectVisibility.PRIVATE);
+          projectDAO.findProjects(request, null, userInfo, VisibilityEnum.Visibility.PRIVATE);
```
```suggestion
          projectDAO.findProjects(request, null, userInfo, ResourceVisibility.PRIVATE);
```
?

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectServiceImpl.java:
```diff
@@ -943,7 +886,8 @@ public void getProjectByName(
       ProjectPaginationDTO projectPaginationDTO =
-          projectDAO.findProjects(findProjects.build(), null, userInfo, ProjectVisibility.PRIVATE);
+          projectDAO.findProjects(
+              findProjects.build(), null, userInfo, VisibilityEnum.Visibility.PRIVATE);
```
```suggestion
              findProjects.build(), null, userInfo, ResourceVisibility.PRIVATE);
```

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectServiceImpl.java:
```diff
@@ -829,7 +771,8 @@ public void getProjects(
       ProjectPaginationDTO projectPaginationDTO =
-          projectDAO.findProjects(findProjects.build(), null, userInfo, ProjectVisibility.PRIVATE);
+          projectDAO.findProjects(
+              findProjects.build(), null, userInfo, VisibilityEnum.Visibility.PRIVATE);
```
```suggestion
              findProjects.build(), null, userInfo, ResourceVisibility.PRIVATE);
```

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -1405,7 +1232,7 @@ public Project setProjectWorkspace(String projectId, WorkspaceDTO workspaceDTO)
               null,
               new CollaboratorUser(authService, currentLoginUserInfo),
-              ProjectVisibility.PRIVATE,
+              VisibilityEnum.Visibility.PRIVATE,
```
```suggestion
              ResourceVisibility.PRIVATE,
```
?

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -1114,7 +976,7 @@ public ProjectPaginationDTO findProjects(
         }
       } else {
-        if (projectVisibility.equals(ProjectVisibility.PRIVATE)) {
+        if (visibility.equals(VisibilityEnum.Visibility.PRIVATE)) {
```
```suggestion
        if (visibility.equals(ResourceVisibility.PRIVATE)) {
```
?

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -1048,15 +910,15 @@ public ProjectPaginationDTO findProjects(
       CollaboratorBase host,
       UserInfo currentLoginUserInfo,
-      ProjectVisibility projectVisibility)
+      VisibilityEnum.Visibility visibility)
```
```suggestion
      ResourceVisibility visibility)
```
?

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -911,7 +871,7 @@ public Project setProjectShortName(String projectId, String projectShortName, Us
             hostCollaboratorBase,
             currentLoginUserInfo,
-            ProjectVisibility.PUBLIC);
+            VisibilityEnum.Visibility.PUBLIC);
```
```suggestion
            ResourceVisibility.PUBLIC);
```
?

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -751,14 +711,14 @@ public Project deepCopyProjectForUser(String srcProjectID, UserInfo newOwner)
         ModelDBConstants.ROLE_PROJECT_ADMIN,
         ModelDBServiceResourceTypes.PROJECT,
-        projectVisibility.equals(ProjectVisibility.ORG_SCOPED_PUBLIC),
+        visibility.equals(VisibilityEnum.Visibility.ORG_SCOPED_PUBLIC),
```
```suggestion
        visibility.equals(ResourceVisibility.ORG_DEFAULT),
```
Due to some miscommunication, we ended up with
ORG_SCOPED_PUBLIC -> ORG_DEFAULT
and
ORG_CUSTOM == ORG_SCOPED_PUBLIC (we should change the proto enum name *very* soon)

coryverta replied:
What is the correct value to be using here?

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -751,14 +711,14 @@ public Project deepCopyProjectForUser(String srcProjectID, UserInfo newOwner)
       WorkspaceType workspaceType,
       String projectId,
-      ProjectVisibility projectVisibility) {
+      VisibilityEnum.Visibility visibility) {
```
```suggestion
      ResourceVisibility visibility) {
```

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -584,7 +544,7 @@ public ProjectPaginationDTO getProjects(
             .build();
     ProjectPaginationDTO projectPaginationDTO =
-        findProjects(findProjects, null, userInfo, ProjectVisibility.PRIVATE);
+        findProjects(findProjects, null, userInfo, VisibilityEnum.Visibility.PRIVATE);
```
```suggestion
        findProjects(findProjects, null, userInfo, ResourceVisibility.PRIVATE);
```

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -557,7 +517,7 @@ public ProjectPaginationDTO getProjects(
       Boolean order,
       String sortKey,
-      ProjectVisibility projectVisibility)
+      VisibilityEnum.Visibility visibility)
```
```suggestion
      ResourceVisibility visibility)
```

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAORdbImpl.java:
```diff
@@ -236,39 +219,16 @@ public Project insertProject(Project project, UserInfo userInfo)
-    }
+        ownerId,
+        ModelDBServiceResourceTypes.PROJECT,
+        CollaboratorTypeEnum.CollaboratorType.READ_WRITE,
```
Why are we defaulting to RW? I'm not sure where this method is called. Should we replace it at the source to have more context?

coryverta replied:
This is ONLY used in the method above (insertProject) to set the initial permissions.  I can set this value to whatever makes sense for a new project.

coryverta replied:
I inlined this call to eliminate the extra method overhead, and now it lets UAC handle the owner ID assignment and defaults to READ_ONLY

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAO.java:
```diff
@@ -278,7 +266,7 @@ ProjectPaginationDTO findProjects(
       CollaboratorBase host,
       UserInfo currentLoginUserInfo,
-      ProjectVisibility projectVisibility)
+      VisibilityEnum.Visibility projectVisibility)
```
```suggestion
      ResourceVisibility projectVisibility)
```

conradoverta commented on backend/src/main/java/ai/verta/modeldb/project/ProjectDAO.java:
```diff
@@ -117,7 +116,7 @@ ProjectPaginationDTO getProjects(
       Boolean order,
       String sortKey,
-      ProjectVisibility projectVisibility)
+      VisibilityEnum.Visibility projectVisibility)
```
```suggestion
      ResourceVisibility projectVisibility)
```

coryverta committed:
Merge branch 'master' into cj/VR-8152/project-creation-uses-setResources

coryverta committed:
Remove support for PUBLIC visibility in RoleService.getReadOnlyAccessibleResourceIds

coryverta committed:
Update backend/src/main/java/ai/verta/modeldb/project/ProjectDAO.java

Co-authored-by: conradoverta <53195685+conradoverta@users.noreply.github.com>

coryverta committed:
Update backend/src/main/java/ai/verta/modeldb/utils/RdbmsUtils.java

Co-authored-by: conradoverta <53195685+conradoverta@users.noreply.github.com>

coryverta committed:
Update backend/src/main/java/ai/verta/modeldb/project/ProjectServiceImpl.java

Co-authored-by: conradoverta <53195685+conradoverta@users.noreply.github.com>

coryverta committed:
Merge conflicts

Commit:
https://jenkins.dev.verta.ai/job/build/job/autoformat/job/modeldb-backend/7/

coryverta committed:
Code formatter

coryverta committed:
Merge remote-tracking branch 'origin/cj/VR-8152/project-creation-uses-setResources' into cj/VR-8152/project-creation-uses-setResources

coryverta committed:
Inline the call to createWorkspacePermissions in insertProject

Default the ownerId to empty so UAC populates it.
Set the default collab type to READ_ONLY

Commit:
https://jenkins.dev.verta.ai/job/build/job/autoformat/job/modeldb-backend/10/

coryverta reviewed (commented):
https://github.com/VertaAI/modeldb/commit/f2d0f8588270716432d112a5ef3ecbf154224833#diff-c8d019d1d18aa75e6ffdc3d9eee30bdda569c52ea122ddfbdbe326693460ec56R1130-R1137 is the original declaration which took only the workspaceServiceId and did the service call.
